### PR TITLE
Fix: update merged probability computation to follow Bayes' rule

### LINKED ISSUE #5152: Computation of merged probabilities for GOLD
Hi,

I'm looking into GOLD for logit distillation and believe there is a bug in the way that the merged probability of tokens in a group is being computed. See the function [here](https://github.com/huggingface/trl/blame/fa313fd57244008953753047795c954a782f9cfc/trl/experimental/gold/gold_trainer.py#L524). Extracting the marginal of the first position and multiplying by the conditionals of the subsequent tokens only results in the correct probability for the actual sequence, as mentioned in the comments. Couldn't this be fixed by applying Bayes' rule?

For instance, assume a group consists of two tokens `t1 = "a"` and `t2 = "b"` at the first and second positions, respectively. To get the distribution over the vocabulary, instead of doing `P(t1 | context) * P(t2 = "b" | t1 = "a", context)` (treating `t1` as variable), which is only correct when `t1 = "a"` and incorrect otherwise, shouldn't we compute `P(t1 = "a" | context) * P(t2 | t1 = "a", context)` (treating `t2` as variable)? This would be correct for all values of `t2`.

### PR DESCRIPTION
**Pull request by vman049**
# What does this PR do?

Fixes the merged-probability computation in the experimental GOLD trainer's `ULDLoss` so it follows the probability chain rule (Bayes' rule) correctly.

Two related corrections in `trl/experimental/gold/gold_trainer.py`:

1. **Off-by-one in answer-logit extraction.** Logits are now sliced starting one position earlier (`start - 1 : start + size - 1`) so that `probs[k]` is the distribution that predicts `token_ids[k]`. Previously the logits were misaligned with the tokens they were meant to predict.

2. **Correct group merging via the chain rule.** When merging tokens aligned to positions `[i, …, i+k]`, the merged distribution now uses the **last position's full distribution** (conditioned on the actual generated prefix tokens) multiplied by the scalar probabilities of the actual earlier tokens:

   ```
   P_merged(y | x) = P(token_i | x) × … × P(token_{i+k-1} | …, x) × P(y | token_i, …, token_{i+k-1}, x)
   ```

   The previous implementation started from the **marginal** distribution at the *first* position and multiplied by the later conditionals, which produced a known bias for counterfactual tokens. Conditioning the full distribution on the actual prefix is valid for every vocabulary entry. The merged distribution remains unnormalized but now preserves correct relative probabilities by the chain rule. More details in the [issue link](https://github.com/huggingface/trl/issues/5152).

`tests/experimental/test_gold_trainer.py` is updated to match the corrected behavior.

Fixes #5152

## Before submitting

- [ ] This PR fixes a typo or improves the docs (you can dismiss the other checks if that's the case).
- [x] Did you read the [contributor guideline](https://github.com/huggingface/trl/blob/main/CONTRIBUTING.md#create-a-pull-request), Pull Request section?
- [x] Was this discussed/approved via a GitHub issue? Please add a link to it if that's the case. → #5152
- [ ] Did you make sure to update the documentation with your changes?
- [x] Did you write any new necessary tests? (updated `tests/experimental/test_gold_trainer.py`)

## AI writing disclosure

- [ ] No AI usage: the PR was written entirely by a human.
- [x] AI-assisted: some parts were suggested or improved by AI, but the PR was written and reviewed by a human.
- [ ] AI-generated: the PR was mostly or fully generated by an AI tool.

## Who can review?

@cmpatino , @qgallouedec 

---

> [!NOTE]
> **Medium Risk**
> Changes experimental GOLD/ULD distillation math and training loss when `bayesian` is selected; default `observed` limits blast radius but misconfiguration could alter distillation behavior.
> 
> **Overview**
> Adds **`uld_token_merge_strategy`** to GOLD (`"observed"` default, `"bayesian"` optional) with config docs and validation.
> 
> **`ULDLoss`** now branches on that setting. **`"bayesian"`** slices answer logits one step earlier so `probs[k]` predicts `token_ids[k]`, and merges split-token groups using the **last** position’s full distribution times scalar probs of the **earlier** actual tokens (chain rule). **`"observed"`** keeps the prior behavior: answer logits at label positions and merge from the **first** position’s marginal times later scalar conditionals.
> 
> For **`"bayesian"`**, answer spans starting at index 0 (e.g. after front truncation) skip the unscorable leading token(s)—by dropping the first aligned group on the extended path or one positional target—avoiding negative-index wrap and yielding zero loss when nothing remains.
> 
> Tests cover both merge formulas, shifted distillation loss vs observed, and zero-start edge cases on positional and extended ULD paths.
> 
> <sup>Reviewed by [Cursor Bugbot](https://cursor.com/bugbot) for commit 110cded26ee1a27f529f41f38a69e1988a6f9bcc. Bugbot is set up for automated code reviews on this repo. Configure [here](https://www.cursor.com/dashboard/bugbot).</sup>

### CODE REVIEW

File: trl/experimental/gold/gold_trainer.py [hunk 1]
```diff
@@ -359,9 +359,9 @@ def _compute_distillation_loss(
                 distillation_losses.append(loss_i)
                 continue
 
-            # Extract answer logits
-            student_answer_logits = student_logits[i, student_start : student_start + student_size]
-            teacher_answer_logits = teacher_logits[i, teacher_start : teacher_start + teacher_size]
+            # Extract answer logits (start one position earlier so probs[k] predicts token_ids[k])
+            student_answer_logits = student_logits[i, student_start - 1 : student_start + student_size - 1]
+            teacher_answer_logits = teacher_logits[i, teacher_start - 1 : teacher_start + teacher_size - 1]
```
[the remark this answers is not in the document]
Comment (vman049) [reply trl/experimental/gold/gold_trainer.py]: `student_start` is the first completion token position, and there is always a prompt before it, so it is always >= 1.

**Comment by vman049** [comment]
@cmpatino / @qgallouedec , would you be able to take a look at this PR when you get a chance?

**Comment by kashif** [comment]
i can also review

**Comment by cmpatino** [comment]
Thank you for uploading the PR! The changes look good, but I’d prefer to keep the new merging method alongside the current one for now. That lets us battle-test it before making it the default.

What do you think about introducing a parameter, `uld_token_merge_strategy`, in [`gold_config.py`](https://github.com/huggingface/trl/blob/main/trl/experimental/gold/gold_config.py)? It could accept two values:
- `"observed"`: current behavior
- `"bayesian"`: the behavior introduced in this PR

I’d keep `"observed"` as the default for now.

This setup would let us progressively sunset the observed merge strategy and migrate to your Bayesian implementation without breaking results for anyone currently using the trainer.

**Comment by vman049** [comment]
Yes, for sure! Sorry for the late reply. Will take care of this today.

**Comment by vman049** [comment]
Apologies for the late follow-up! @cmpatino , could you please take a look at the revision when you get a chance?

File: trl/experimental/gold/gold_trainer.py [hunk 2]
```diff
@@ -404,9 +405,13 @@ def _compute_distillation_loss(
...
-            # Extract answer logits
-            student_answer_logits = student_logits[i, student_start : student_start + student_size]
-            teacher_answer_logits = teacher_logits[i, teacher_start : teacher_start + teacher_size]
+            # Extract answer logits. "bayesian" starts one position earlier so probs[k] predicts token_ids[k].
+            if self.token_merge_strategy == "bayesian":
+                student_answer_logits = student_logits[i, student_start - 1 : student_start + student_size - 1]
+                teacher_answer_logits = teacher_logits[i, teacher_start - 1 : teacher_start + teacher_size - 1]
+            else:
+                student_answer_logits = student_logits[i, student_start : student_start + student_size]
+                teacher_answer_logits = teacher_logits[i, teacher_start : teacher_start + teacher_size]
```
[the remark this answers is not in the document]
Comment (vman049) [reply trl/experimental/gold/gold_trainer.py:428]: Same as before: `student_start` is the first completion token position, and there is always a prompt before it, so it is always >= 1.

File: trl/experimental/gold/gold_trainer.py [hunk 3]
```diff
@@ -543,24 +544,21 @@ def _merge_probabilities_with_alignment_groups(self, probs, alignment_groups, to
                         "This is required for mathematically correct probability merging."
                     )
```
[the remark this answers is not in the document]
Comment (vman049) [reply trl/experimental/gold/gold_trainer.py:559]: Reworded the message so it no longer claims "mathematically correct"

File: trl/experimental/gold/gold_config.py [hunk 4]
```diff
@@ -79,6 +79,12 @@ class GOLDConfig(SFTConfig):
             across student and teacher tokenizations. When `True`, the trainer will compute token mappings and merge
             probabilities for split tokens; when `False`, ULD will use simple positional truncation like in the
             original ULD paper.
+        uld_token_merge_strategy (`str`, *optional*, defaults to `"observed"`):
+            Strategy used to merge token probabilities across split tokens when extended ULD alignment is enabled.
+            Either `"observed"` (multiply the marginal distribution at the first position by the scalar conditional
+            probabilities of the actual later tokens) or `"bayesian"` (use the last position's full distribution,
+            conditioned on the actual prefix tokens, multiplied by the scalar probabilities of the earlier tokens,
+            following the chain rule).
```
[the remark this answers is not in the document]
Comment (vman049) [reply trl/experimental/gold/gold_config.py]: Updated both the class docstring and the field `help` for `uld_token_merge_strategy` to document that `"bayesian"` also shifts the answer-logit slice one position earlier (so `probs[k]` predicts `token_ids[k]`), and that this shift applies regardless of `use_extended_uld`.

File: tests/experimental/test_gold_trainer.py [hunk 5]
```diff
@@ -892,10 +893,31 @@ def test_merge_probabilities_multiplies_split_tokens():
...
+    token_ids = [0, 1]  # Actual generated tokens
+
+    merged = loss._merge_probabilities_with_alignment_groups(probs, [[0, 1]], token_ids=token_ids)
+
+    # Expected: last position's full distribution × scalar prob of actual token at earlier position
+    # P_merged(y) = probs[1](y) × probs[0, token_ids[0]] = probs[1] × probs[0, 0]
+    expected = probs[1] * probs[0, 0]  # probs[0, 0] = 0.6
+    # Expected unnormalized: [0.2 * 0.6, 0.5 * 0.6, 0.3 * 0.6] = [0.12, 0.30, 0.18]
 
     torch.testing.assert_close(merged[0], expected)
```
[the remark this answers is not in the document]
Comment (vman049) [reply tests/experimental/test_gold_trainer.py:922]: Added `test_compute_distillation_loss_bayesian_shifts_answer_logits`, which exercises the positional path (`use_extended_uld=False`) and asserts the loss is computed from the shifted slice (positions `[0, 1]`), with a check that `"observed"` uses the unshifted positions `[1, 2]`.

**Comment by vman049** [comment]
@kashif / @cmpatino , could you please take a look at the latest revisions?

File: trl/experimental/gold/gold_trainer.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (cmpatino) [reply trl/experimental/gold/gold_trainer.py:428]: I agree with Copilot in this one.

I think the `start == 0` case is still reachable, even though the original example always has a prompt before the completion.

The issue is GOLD’s front-truncation path. In `_prepare_dataset_with_original_text`, when the rendered sequence is longer than `max_length`, we keep the last `max_length` tokens:

```python
drop = len(input_ids) - max_length
input_ids = input_ids[drop:]
byte_offsets = byte_offsets[drop:]
completion_start = max(0, completion_start - drop)
```

We've seen this happen when we use long training rollouts.

So if truncation drops the full prompt and eats into the completion, `completion_start` becomes 0. The test `test_prepared_tokenized_rows_rebase_byte_offsets_when_truncation_eats_into_completion` addresses that scenario:

```python
assert row["completion_mask"] == [1] * max_length
```

Once that row goes through the collator, all labels can start at index `0`, and `_get_start_and_size_answers(...)` can return `start = 0`. In `"bayesian"` mode, `student_start - 1` then becomes `-1`, so the shifted logit slice is empty or wrong.

I don’t think `max(start - 1, 0)` is the right fix, because the first completion token genuinely has no causal predictor logit left in the truncated sequence. Instead, we should skip the unscorable leading span.

What I would suggest:
- For `use_extended_uld=True`, build the full byte-offset alignment first. If either `student_start == 0` or `teacher_start == 0`, drop the first aligned group pair from both sides, then advance `student_start` and `teacher_start` by the number of tokens in the dropped groups. This keeps student/teacher aligned to the same text span.
- For `use_extended_uld=False`, if either side starts at `0`, drop the first positional target from both sides by advancing both starts by `1` and decrementing both sizes by `1`.
- If nothing remains after dropping, append the existing zero-loss value for that sample and continue.

**Review by cmpatino — COMMENTED** [review]
Everything LGTM except for the comment about handling the case when `start=0`.

Thank you for pushing this effort!

**Comment by vman049** [comment]
Thanks, @cmpatino, for the detailed feedback! Took your suggestion. Latest commit skips the unscorable leading span in "bayesian" mode when an answer starts at index 0.

**Review by cmpatino — DISMISSED** [review]
LGTM!

Thank you for leading this work from the initial discussion (https://github.com/huggingface/trl/issues/5152) through to the implementation.

Very nice addition to the trainer, and super excited to try it out.

**Comment by cmpatino** [comment]
@vman049, the changes look good, but can you please check that all the tests are green before merging?

I'm seeing an error in `Tests (experimental) / Check code quality (pull_request)`, which is a tiny formatting issue that can be solved quickly with 

```
make precommit
``` 

as described in [here](https://github.com/huggingface/trl?tab=contributing-ov-file#submitting-a-pull-request-pr).

~For the failure on `[Tests (experimental) / Tests (experimental) (pull_request)]`, it's an issue with another trainer. I'm pushing a fix right now.~ This is now fixed

File: trl/experimental/gold/gold_trainer.py [hunk 6]
```diff
@@ -404,9 +405,37 @@ def _compute_distillation_loss(
...
+                        distillation_losses.append(student_logits[i].sum() * 0.0)
+                        continue
+                    # Drop the first aligned group pair (advance by the tokens it covered).
+                    student_drop, teacher_drop = len(student_groups[0]), len(teacher_groups[0])
+                else:
+                    # Drop the first positional target from both sides.
+                    student_drop = teacher_drop = 1
+
+                student_start, student_size = student_start + student_drop, student_size - student_drop
+                teacher_start, teacher_size = teacher_start + teacher_drop, teacher_size - teacher_drop
```
[the remark this answers is not in the document]
Comment (vman049) [reply trl/experimental/gold/gold_trainer.py:417]: Loss compares student/teacher element-wise per pair, so dropping the unscorable leading token on both sides is required to keep the pairs aligned

File: trl/experimental/gold/gold_trainer.py [hunk 7]
```diff
@@ -404,9 +405,37 @@ def _compute_distillation_loss(
...
+                    if student_byte_offsets is None or teacher_byte_offsets is None:
+                        raise ValueError("Byte offsets are required when `use_extended_uld=True`.")
+                    s_answer = student_byte_offsets[i, student_start : student_start + student_size].tolist()
+                    t_answer = teacher_byte_offsets[i, teacher_start : teacher_start + teacher_size].tolist()
+                    student_groups, teacher_groups = self._align_by_byte_offsets(s_answer, t_answer)
+                    if not student_groups or not teacher_groups:
+                        distillation_losses.append(student_logits[i].sum() * 0.0)
+                        continue
+                    # Drop the first aligned group pair (advance by the tokens it covered).
+                    student_drop, teacher_drop = len(student_groups[0]), len(teacher_groups[0])
```
[the remark this answers is not in the document]
Comment (vman049) [reply trl/experimental/gold/gold_trainer.py:411]: Bayesian merge folds the whole byte-group into one chain-rule distribution that needs the first token's prob (missing when `start=0`), so the entire group is unscorable. Dropping the aligned pair keeps the sides byte-aligned

**Comment by vman049** [comment]
Thanks, @cmpatino ! Yes, did not run `make precommit` the first time, but it's done now. Formatting should be fixed. Have also responded to other two comments (although they're already marked as resolved). Please let me know if anything else is needed.

**Review by cmpatino — APPROVED** [review]
LGTM and all the tests are passing. 

Merging it now. Thanks for the contribution, @vman049!
